packages/network/src/logic/proxy/ProxyStreamConnectionClient.ts

```diff
@@ -0,0 +1,289 @@
+            logger.warn(reason)
+            throw reason
+        }
+        logger.info(`Open proxy connection to ${targetNodeId} on ${streamPartId}`)
+        if (!this.connections.has(streamPartId)) {
+            this.connections.set(streamPartId, new Map())
+        }
+        this.connections.get(streamPartId)!.set(targetNodeId, direction)
+        try {
+            await this.connectAndHandshake(streamPartId, targetNodeId, direction, userId)
+        } catch (err) {
+            logger.warn(`Failed to create a proxy ${direction} stream connection to ${targetNodeId} for stream ${streamPartId}:\n${err}`)
+            this.removeConnection(streamPartId, targetNodeId)
+            this.emit(Event.CONNECTION_REJECTED, targetNodeId, streamPartId, direction, err)
+        } finally {
+            const trackerId = this.trackerManager.getTrackerId(streamPartId)
+            this.trackerManager.disconnectFromSignallingOnlyTracker(trackerId)
+        }
+    }
+
+    private async connectAndHandshake(streamPartId: StreamPartID, targetNodeId: NodeId, direction: ProxyDirection, userId: string): Promise<void> {
+        const trackerId = this.trackerManager.getTrackerId(streamPartId)
+        const trackerAddress = this.trackerManager.getTrackerAddress(streamPartId)
+        await this.trackerManager.connectToSignallingOnlyTracker(trackerId, trackerAddress)
+        await withTimeout(this.nodeToNode.connectToNode(targetNodeId, trackerId, false), this.nodeConnectTimeout)
+        await this.nodeToNode.requestProxyConnection(targetNodeId, streamPartId, direction, userId)
+    }
+
+    private async closeRandomConnections(streamPartId: StreamPartID, connectionCount: number): Promise<void> {
+        logger.debug(`Close ${connectionCount} random connections on ${streamPartId}`)
+        const proxiesToDisconnect = shuffle(Array.from(this.getConnections(streamPartId).keys()))
+            .splice(0, connectionCount)
+        await Promise.allSettled(proxiesToDisconnect.map((node) => this.closeConnection(streamPartId, node)))
+    }
+
+    private async closeConnection(streamPartId: StreamPartID, targetNodeId: NodeId): Promise<void> {
+        if (this.getConnections(streamPartId).has(targetNodeId)
+            && this.streamPartManager.hasOnewayConnection(streamPartId, targetNodeId)
+        ) {
+            logger.info(`Close proxy connection to ${targetNodeId} on ${streamPartId}`)
+            await this.nodeToNode.leaveStreamOnNode(targetNodeId, streamPartId)
+            this.node.emit(NodeEvent.ONE_WAY_CONNECTION_CLOSED, targetNodeId, streamPartId)
+            this.removeConnection(streamPartId, targetNodeId)
+        }
+    }
+
+    private getConnections(streamPartId: StreamPartID): Map<NodeId, ProxyDirection> {
+        return this.connections.get(streamPartId) ?? new Map()
+    }
+
+    private hasConnection(nodeId: NodeId, streamPartId: StreamPartID): boolean {
+        if (!this.definitions.has(streamPartId)) {
```

Comment:
Would it make sense to remove this check? Then hasConnection and getConnections would have the same logic, i.e. just analyzing `this.connections` field.